File: LeetCode/cpp/1022.cpp
//Link: https://leetcode.com/problems/sum-of-root-to-leaf-binary-numbers/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        return solve(root, 0);
    }
    int solve(TreeNode* root, int value) {
        if(!root)
            return 0;
        if(!root->left && !root->right)
            return (value << 1) + root->val;
        value = (value << 1) + root->val;
        return solve(root->left, value) + solve(root->right, value);
    }
};

File: LeetCode/py/1022.py
#Link: https://leetcode.com/problems/sum-of-root-to-leaf-binary-numbers/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, root: TreeNode) -> int:
        return self.solve(root, 0)
    def solve(self, root: TreeNode, value: int) -> int:
        if root == None:
            return 0
        if root.left == None and root.right == None:
            return (value << 1) + root.val;
        value = (value << 1) + root.val
        return self.solve(root.left, value) + self.solve(root.right, value)
